### PR TITLE
Fetch all table/view names across schemas with one Hive metastore call (when Hive views enabled)

### PR DESCRIPTION
**Pull request by huberty89**
## Description
This PR replaces https://github.com/trinodb/trino/pull/16847 but drops last commits because of a regression - I will try to reintroduce that with a fix in separate PR.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java [hunk 1]
```diff
@@ -79,7 +79,7 @@
     private static final InformationSchemaColumnHandle TABLE_NAME_COLUMN_HANDLE = new InformationSchemaColumnHandle("table_name");
     private static final InformationSchemaColumnHandle ROLE_NAME_COLUMN_HANDLE = new InformationSchemaColumnHandle("role_name");
     private static final InformationSchemaColumnHandle GRANTEE_COLUMN_HANDLE = new InformationSchemaColumnHandle("grantee");
-    private static final int MAX_PREFIXES_COUNT = 100;
+    private static final int MAX_PREFIXES_COUNT = 4;
```
[the remark this answers is not in the document]
Comment (huberty89) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java]: Accident, reverted

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreClient.java [hunk 2]
```diff
@@ -47,6 +49,12 @@ List<String> getAllDatabases()
     Database getDatabase(String databaseName)
             throws TException;
 
+    Optional<List<SchemaTableName>> getAllTables()
+            throws TException;
+
+    Optional<List<SchemaTableName>> getAllViews()
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreClient.java]: Update methods order to be coherent with `ThriftMetastore`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 3]
```diff
@@ -117,10 +117,12 @@
     private final LoadingCache<String, List<String>> databaseNamesCache;
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:117]: Added few tests

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java [hunk 4]
```diff
@@ -177,13 +186,32 @@ public Database getDatabase(String dbName)
         return client.getDatabase(dbName);
     }
 
+    private static Optional<List<SchemaTableName>> parseTableMetadata(List<TableMeta> tablesMetadata)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java:189]: nit: move this method below the usage

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 5]
```diff
@@ -248,6 +254,27 @@ public void testGetAllTable()
         assertEquals(metastore.getTableNamesStats().getHitRate(), 1.0 / 3);
     }
 
+    @Test
+    public void testBatchGetAllTable()
+    {
+        assertEquals(mockClient.getAccessCount(), 0);
+        assertEquals(metastore.getAllTables(), Optional.of(ImmutableList.of(new SchemaTableName(TEST_DATABASE, TEST_TABLE))));
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:261]: nit. Extract variable for schema table name

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 6]
```diff
@@ -1076,6 +1103,70 @@ public void testDropTable()
...
+        assertThat(metastore.getAllDatabases()).containsExactly(TEST_DATABASE);
+        assertThat(mockClient.getAccessCount()).isEqualTo(3); // should read it from cache
+    }
+
+    @Test
+    public void testAllTables()
+    {
+        assertThat(mockClient.getAccessCount()).isEqualTo(0);
+
+        assertThat(metastore.getAllTables()).contains(ImmutableList.of(new SchemaTableName(TEST_DATABASE, TEST_TABLE)));
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:1141]: nit: extract variable for schema table name

**Review by kokosing — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java (hunk 4 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java:189]: Done

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/ForwardingHiveMetastore.java [hunk 7]
```diff
@@ -126,6 +126,18 @@ public List<String> getAllViews(String databaseName)
...
     }
 
+    @Override
+    public Optional<List<SchemaTableName>> getAllTables()
+    {
+        return delegate.getAllTables();
+    }
+
+    @Override
+    public Optional<List<SchemaTableName>> getAllViews()
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/ForwardingHiveMetastore.java:136]: Fixed

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java [hunk 8]
```diff
@@ -196,6 +224,18 @@ public List<String> getAllViews(String databaseName)
...
 
+    @Override
+    public Optional<List<SchemaTableName>> getAllViews()
+            throws TException
+    {
+        return alternativeCall(
+                exception -> !isUnknownMethodExceptionalResponse(exception),
+                chosenGetAllViewsAlternative,
+                () -> parseTableMetadata(client.getTableMeta("*", "*", ImmutableList.of(VIRTUAL_VIEW.name()))),
+                // fallback to fetching views per schema if getTableMeta call is not available
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java]: Ok, removed

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java (hunk 5 above)
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java]: Extracted

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java (hunk 6 above)
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java]: Extracted

**Comment by huberty89** [comment]
@findepi Could you take a look?

**Comment by findepi** [comment]
discussed offline

**Comment by huberty89** [comment]
I feel I need to run `TestHiveMetastoreMetadataQueriesAccessOperations` for both cases:
- `HiveMetastore` with implemented `getAllTables()` and `getAllViews()` - so this will cover `ThriftMetastore`
- the backward compatible test - so those changes does not brake/slowdown for example `GlueHiveMetastore`

WDYT? @kokosing @Praveen2112 @findepi

**Comment by huberty89** [comment]
Discussed offline

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java [hunk 9]
```diff
@@ -196,6 +217,24 @@ public List<String> getAllViews(String databaseName)
...
+            throws TException
+    {
+        return alternativeCall(
+                exception -> !isUnknownMethodExceptionalResponse(exception),
+                chosenGetAllViewsAlternative,
+                () -> parseTableMetadata(client.getTableMeta("*", "*", ImmutableList.of(VIRTUAL_VIEW.name()))),
+                Optional::empty);
+    }
+
+    private static Optional<List<SchemaTableName>> parseTableMetadata(List<TableMeta> tablesMetadata)
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java:231]: nit : Can we have different method name - `getSchemaTableNames` or something like that.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 10]
```diff
@@ -1117,6 +1145,70 @@ public void testDropTable()
         assertEquals(mockClient.getAccessCount(), 5);
     }
 
+    @Test
+    public void testAllDatabases()
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:1149]: Can we extract it as a different commit.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java [hunk 11]
```diff
@@ -326,6 +326,20 @@ public synchronized List<String> getAllViews(String databaseName)
         return tables.build();
     }
 
+    @Override
+    public synchronized Optional<List<SchemaTableName>> getAllTables()
+    {
+        return Optional.of(this.relations.keySet().stream()
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java:332]: nit: Optional.of(ImmutableList.copyOf(relations.keySet()))` and for view also

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/MockThriftMetastoreClient.java [hunk 12]
```diff
@@ -384,13 +402,13 @@ private static Partition getPartitionsByNamesUnchecked(String name)
     @Override
     public void createDatabase(Database database)
     {
-        throw new UnsupportedOperationException();
+        // No-op, make sure the cache invalidation logic in CachingHiveMetastore will be passed through
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/MockThriftMetastoreClient.java:405]: Is it for the `testAllDatabases` ?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java [hunk 13]
```diff
@@ -67,6 +72,11 @@
         private static final List<String> TABLES_PER_SCHEMA = IntStream.range(0, TEST_TABLES_IN_SCHEMA_COUNT)
                 .mapToObj("test_table_%d"::formatted)
                 .collect(toImmutableList());
+        private static final ImmutableList<SchemaTableName> ALL_TABLES = SCHEMAS.stream()
+                .flatMap(schema -> TABLES_PER_SCHEMA.stream()
+                        .map(table -> new SchemaTableName(schema, table)))
+                .collect(toImmutableList());
+        private boolean allTablesViewsImplemented;
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java:79]: nit : A new line seperation.

**Review by Praveen2112 — COMMENTED** [review]
Looks good. Need to run another round of review on `TestHiveMetastoreMetadataQueriesAccessOperations`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/MockThriftMetastoreClient.java (hunk 12 above)
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/MockThriftMetastoreClient.java:405]: Yes it is

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java [hunk 14]
```diff
@@ -59,6 +58,59 @@
...
+        @Override
+        public List<String> getAllViews(String databaseName)
+        {
+            return ImmutableList.of();
+        }
+
+        @Override
+        public Optional<Table> getTable(String databaseName, String tableName)
+        {
+            if (!SCHEMAS.contains(databaseName) || !TABLES_PER_SCHEMA.contains(tableName)) {
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java]: Ok, I checked and yes, this path does not occur, I removed those checks.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java [hunk 15]
```diff
@@ -59,6 +58,59 @@
 
     private CountingAccessHiveMetastore metastore;
 
+    private static class MockHiveMetastore
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java]: Done

[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java]: Moved

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java [hunk 16]
```diff
@@ -116,7 +118,9 @@ protected ThriftMetastoreClient create(TransportSupplier transportSupplier, Stri
                 chosenTableParamAlternative,
                 chosenGetAllViewsPerDatabaseAlternative,
                 chosenAlterTransactionalTableAlternative,
-                chosenAlterPartitionsAlternative);
+                chosenAlterPartitionsAlternative,
+                chosenGetAllTablesAlternative,
+                chosenGetAllViewsAlternative);
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java]: Checked the order

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 17]
```diff
@@ -863,6 +863,53 @@ public List<String> getAllViews(String databaseName)
...
+            throw propagate(e);
+        }
+    }
+
+    @Override
+    public Optional<List<SchemaTableName>> getAllViews()
+    {
+        // Without translateHiveViews, Hive views are represented as tables in Trino,
+        // and they should not be returned from ThriftHiveMetastore.getAllViews() call
+        if (translateHiveViews) {
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java]: Inverted

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java [hunk 18]
```diff
@@ -126,7 +131,9 @@ public ThriftHiveMetastoreClient(
             AtomicInteger chosenTableParamAlternative,
             AtomicInteger chosenGetAllViewsPerDatabaseAlternative,
             AtomicInteger chosenAlterTransactionalTableAlternative,
-            AtomicInteger chosenAlterPartitionsAlternative)
+            AtomicInteger chosenAlterPartitionsAlternative,
+            AtomicInteger chosenGetAllTablesAlternative,
+            AtomicInteger chosenGetAllViewsAlternative)
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java]: Same as https://github.com/trinodb/trino/pull/17127#discussion_r1185866223

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java [hunk 19]
```diff
@@ -184,6 +193,18 @@ public List<String> getAllTables(String databaseName)
...
     }
 
+    @Override
+    public Optional<List<SchemaTableName>> getAllTables()
+            throws TException
+    {
+        return alternativeCall(
+                exception -> !isUnknownMethodExceptionalResponse(exception),
+                chosenGetAllTablesAlternative,
+                // Empty table types argument (the 3rd one) means all types
```
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java]: Added

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java (hunk 9 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java]: Renamed

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java (hunk 11 above)
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java]: Done

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java (hunk 13 above)
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java]: Added

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 20]
```diff
@@ -773,6 +773,17 @@ public Optional<Object> getInfo(ConnectorTableHandle tableHandle)
     @Override
     public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> optionalSchemaName)
     {
+        if (optionalSchemaName.isEmpty()) {
+            Optional<List<SchemaTableName>> allTables = metastore.getAllTables();
+            if (allTables.isPresent()) {
+                return ImmutableList.<SchemaTableName>builder()
+                        .addAll(allTables.orElseThrow().stream()
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:780]: orElseThrow -> get

this provides IDE support (warning if `isPresent` is later removed)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 21]
```diff
@@ -2659,6 +2670,14 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
     @Override
     public List<SchemaTableName> listViews(ConnectorSession session, Optional<String> optionalSchemaName)
     {
+        if (optionalSchemaName.isEmpty()) {
+            Optional<List<SchemaTableName>> allViews = metastore.getAllViews();
+            if (allViews.isPresent()) {
+                return allViews.orElseThrow().stream()
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2676]: orElseThrow -> get

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java [hunk 22]
```diff
@@ -62,10 +62,20 @@ default void updatePartitionStatistics(Table table, String partitionName, Functi
 
     List<String> getAllTables(String databaseName);
 
+    /**
+     * @return List of table names from all schemas or Optional.empty if operation is not supported
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java:66]: Lists tables, views and materialized views

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java [hunk 23]
```diff
@@ -62,10 +62,20 @@ default void updatePartitionStatistics(Table table, String partitionName, Functi
...
+     */
+    Optional<List<SchemaTableName>> getAllTables();
+
     List<String> getTablesWithParameter(String databaseName, String parameterKey, String parameterValue);
 
     List<String> getAllViews(String databaseName);
 
+    /**
+     * @return List of view names from all schemas or Optional.empty if operation is not supported
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java:75]: Can this include materialized views? let's document

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 24]
```diff
@@ -863,6 +863,54 @@ public List<String> getAllViews(String databaseName)
...
+    }
+
+    @Override
+    public Optional<List<SchemaTableName>> getAllViews()
+    {
+        // Without translateHiveViews, Hive views are represented as tables in Trino,
+        // and they should not be returned from ThriftHiveMetastore.getAllViews() call
+        if (!translateHiveViews) {
+            return Optional.empty();
+        }
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:894]: we could remove this IF and adjust the code elsewhere.
We should follow-up on this.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java [hunk 25]
```diff
@@ -326,6 +326,18 @@ public synchronized List<String> getAllViews(String databaseName)
         return tables.build();
     }
 
+    @Override
+    public synchronized Optional<List<SchemaTableName>> getAllTables()
+    {
+        return Optional.of(ImmutableList.copyOf(relations.keySet()));
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/InMemoryThriftMetastore.java:332]: it includes views, right?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java [hunk 26]
```diff
@@ -149,21 +183,36 @@ public void testSelectSchemasWithLikeOverSchemaName()
         assertMetastoreInvocations("SELECT * FROM system.jdbc.schemas WHERE table_schem LIKE 'test%'", ImmutableMultiset.of(GET_ALL_DATABASES));
     }
 
-    @Test
-    public void testSelectTablesWithoutPredicate()
+    @Test(dataProviderClass = DataProviders.class, dataProvider = "trueFalse")
+    public void testSelectTablesWithoutPredicate(boolean allTablesViewsImplemented)
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java:106]: i'd rather go opposite direction, to ensure test coverage stays same for both cases

```
assertMetastoreInvocations( <query>
     allTablesViewsImplemented
             ? <expected if true>
             : <expected if false>
```

**Comment by findepi** [comment]
the fact that TestHiveMetastoreMetadataQueriesAccessOperations.java is refactored in the commit introducing the improvement makes it hard to see what is the effect of the improvement on the invocation counts.

let's maybe introduce the dataprovider in TestHiveMetastoreMetadataQueriesAccessOperations.java as a prep commit (with some TODO / commit comment saying this is just a preparation step)

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java:106]: Applied @findepi  suggestion

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 20 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Done

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 21 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Done

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java (hunk 22 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java]: Done

**Comment by huberty89** [comment]
@findepi discussed offline: about filtering material views from views - this PR is only about performance improvement, I would not change current behavior even if it not correct - we can do it as a followup
cc @kokosing

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java (hunk 23 above)
Comment (huberty89) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastore.java]: Updated

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 27]
```diff
@@ -773,6 +773,17 @@ public Optional<Object> getInfo(ConnectorTableHandle tableHandle)
     @Override
     public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> optionalSchemaName)
     {
+        if (optionalSchemaName.isEmpty()) {
+            Optional<List<SchemaTableName>> allTables = metastore.getAllTables();
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:779]: this is later filtered out
```
!isHiveSystemSchema(table.getSchemaName()))
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java [hunk 28]
```diff
@@ -100,12 +105,12 @@ public void testSelectSchemasWithLikeOverSchemaName()
     @Test(dataProviderClass = DataProviders.class, dataProvider = "trueFalse")
     public void testSelectTablesWithoutPredicate(boolean allTablesViewsImplemented)
     {
+        mockMetastore.setAllTablesViewsImplemented(allTablesViewsImplemented);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestHiveMetastoreMetadataQueriesAccessOperations.java:108]: since `mockMetastore` is created once, there should be a `@BeforeMethod` that resets this, so tests fully isolated

**Review by findepi — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by mosabua** [comment]
@huberty89 and @findepi we had to invent our own release notes entry since none was suggested. Please provide more clarity in future PRs. We ended up with 

```
## Hive connector

* Improve performance of querying `information_schema.tables` when using the
  Hive metastore. ({issue}`17127`)
```

However we are not clear if this applies to all other connectors that use the Hive metastore and therefore need to add this entry to Delta Lake, Iceberg and Hudi connector sections. 

And we are also not sure about your mention about "when Hive views enabled" .. does that limit the scope to only the Hive connector after all?

**Comment by huberty89** [comment]
@mosabua 

> However we are not clear if this applies to all other connectors that use the Hive metastore and therefore need to add this entry to Delta Lake, Iceberg and Hudi connector sections.

This optimization should applies in every connector which uses and it's configured with HMS, so: Delta, Hive, Hudi, Iceberg

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 29]
```diff
@@ -775,6 +775,17 @@ public Optional<Object> getInfo(ConnectorTableHandle tableHandle)
...
```
Comment (hashhar) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:779]: if `optionalSchemaName` points to a single schema then `metastore.getAllTables` does more work than required, which can lead to slower perf compared to older impl.

cc: @kokosing @huberty89

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java (hunk 24 above)
Comment (hashhar) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:894]: the impl of this method differs from `getAllViews(String databaseName)`. that one seems to return tables with `PRESTO_VIEW_FLAG` set to true when translation is disabled instead of empty result. 

was this intentional? Or am I misreading the code?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 29 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:779]: > if optionalSchemaName points to a single schema

but then we don't enter this code path, do we?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java (hunk 24 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:894]: I think the initial implementation was supposed to work only when translateHiveViews or when !translateHiveViews (not sure which one).

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 29 above)
Comment (hashhar) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:779]: indeed, i should read more carefully. Such an embarrassing mistake. 😄
